### DISCUSSION TITLE
Collection initialization syntax with out/ref parameters

### DISCUSSION THREAD

**Discussion by Andrew-Hanlon** [dotnet/csharplang#3015] [Language Ideas]
I have an existing library that uses the convention-based collection initialization syntax to cleanly handle a mixture of `Add` methods. For a new requirement, I built an `Add` method that requires an `out` parameter. When trying to use this method within the list init syntax, I discovered that this is not supported. This struck me as odd/unintuitive given that the syntax is sugar around direct method calls. 

Contrived example:

````csharp
public class FooList : List<string>
{
    public void Add(bool a, int x){ }    
    
    public void Add(string input, out int x) =>    
        int.TryParse(input, out x);    
}

var list = new FooList 
{
    "Test",
    {true, 5},
    {"5", out var z} // *** syntax error ***
};

// Above is syntactic sugar for:

list = new FooList();
list.Add("Test");
list.Add(true, 5);
list.Add("5", out var y);
````

Are there specific reasons not to support out/ref paramters in collection initialization methods (other than effort/value/prioritization)? Would this be an acceptible addition to the language?

**Comment by CyrusNajmabadi** [comment]
> Are there specific reasons not to support out/ref paramters in collection initialization methods

Given how niche this would be (effectively just your api) it would be making an API to support far too narrow a case.

**Comment by YairHalberstadt** [comment]
> (other than effort/value/prioritization)? Would this be an acceptible addition to the language?

Even if you were to spec and implement this yourself, and it required no review, and you wrote the documentation, etc. the complexity of the implementation would have to live on in the compiler forever.

When you compare that to simply writing your add manually, the decision seems obvious.

However IMO there is a reason, albeit not a strong one. If an Add method returns an out parameter, then it's not strictly an initialisation method. It does more, and so doesn't belong in an object initialiser.

**Comment by huoyaoyuan** [comment]
Sorry for commenting on the wrong issue.
